test/e2e/auth/certificates.go

```diff
@@ -108,17 +154,247 @@ var _ = SIGDescribe("Certificates API", func() {
+
+		signerName := "example.com/e2e-" + f.UniqueName
+		csrTemplate := &certificatesv1beta1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{GenerateName: "e2e-example-csr-"},
+			Spec: certificatesv1beta1.CertificateSigningRequestSpec{
+				Request:    csrData,
+				SignerName: &signerName,
+				Usages:     []certificatesv1beta1.KeyUsage{certificatesv1beta1.UsageDigitalSignature, certificatesv1beta1.UsageKeyEncipherment, certificatesv1beta1.UsageServerAuth},
+			},
+		}
+
+		// Discovery
+
+		ginkgo.By("getting /apis")
+		{
+			discoveryGroups, err := f.ClientSet.Discovery().ServerGroups()
+			framework.ExpectNoError(err)
+			found := false
+			for _, group := range discoveryGroups.Groups {
+				if group.Name == certificatesv1beta1.GroupName {
+					for _, version := range group.Versions {
+						if version.Version == csrVersion {
+							found = true
+							break
+						}
+					}
+				}
+			}
+			framework.ExpectEqual(found, true, fmt.Sprintf("expected certificates API group/version, got %#v", discoveryGroups.Groups))
+		}
+
+		ginkgo.By("getting /apis/certificates.k8s.io")
+		{
+			group := &metav1.APIGroup{}
+			err := f.ClientSet.Discovery().RESTClient().Get().AbsPath("/apis/certificates.k8s.io").Do(context.TODO()).Into(group)
+			framework.ExpectNoError(err)
+			found := false
+			for _, version := range group.Versions {
+				if version.Version == csrVersion {
+					found = true
+					break
+				}
+			}
+			framework.ExpectEqual(found, true, fmt.Sprintf("expected certificates API version, got %#v", group.Versions))
+		}
+
+		ginkgo.By("getting /apis/certificates.k8s.io/" + csrVersion)
+		{
+			resources, err := f.ClientSet.Discovery().ServerResourcesForGroupVersion(certificatesv1beta1.SchemeGroupVersion.String())
+			framework.ExpectNoError(err)
+			foundCSR, foundApproval, foundStatus := false, false, false
+			for _, resource := range resources.APIResources {
+				switch resource.Name {
+				case "certificatesigningrequests":
+					foundCSR = true
+				case "certificatesigningrequests/approval":
+					foundApproval = true
+				case "certificatesigningrequests/status":
+					foundStatus = true
+				}
+			}
+			framework.ExpectEqual(foundCSR, true, fmt.Sprintf("expected certificatesigningrequests, got %#v", resources.APIResources))
+			framework.ExpectEqual(foundApproval, true, fmt.Sprintf("expected certificatesigningrequests/approval, got %#v", resources.APIResources))
+			framework.ExpectEqual(foundStatus, true, fmt.Sprintf("expected certificatesigningrequests/status, got %#v", resources.APIResources))
+		}
+
+		// Main resource create/read/update/watch operations
+
+		ginkgo.By("creating")
+		_, err = csrClient.Create(context.TODO(), csrTemplate, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		_, err = csrClient.Create(context.TODO(), csrTemplate, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		createdCSR, err := csrClient.Create(context.TODO(), csrTemplate, metav1.CreateOptions{})
```

Comment:
CSRs are non-namespaced resources, so these need explicit cleanup handling right? Perhaps create a `createCSRWithCleanupFunc` function which returns a `CleanupFunc`?

Reply:
the deletecollection at the end cleans them up